### DISCUSSION TITLE
app.prepare() keeps on calling thenable code inside of it in custom server

### DISCUSSION THREAD

**Discussion by kvedantmahajan** [vercel/next.js#19439] [Help]
# Bug report

All the code block inside of the app.prepare().then() keeps on calling itself. There is a huge memory therefore consumed in the very beginning to it. This is a bug because I suppose we followed the official doc guidelines from here - https://nextjs.org/docs/advanced-features/custom-server

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. create a `server.js` file along with nextjs setup
2. Create a app instance inside of it. `const server = express();`
3. Add some other code. Say a universal express route to console.log the `req.url`
4. Open Node devtools and see that the heapsize increases exponentially. Notice that the terminal shows repeated console.logs from universal route. 

Code Snippet for *server.js*

```
app.prepare().then(() => {
    const server = express();

    const devProxy = require('./proxy.js');
    const {
        PORT,
        NODE_ENV,
        STATIC_PATH
    } = process.env;

    server.use(bodyParser.json());
    app.setAssetPrefix(STATIC_PATH);

     server.get('*', async (req, res) => {
        console.log(req.url, 'render url');
        return handler(req, res);
     });
});
```
The moment I comment out the app.prepare block the app sure falls back to normal heap memory but it doesn't loads up any route at all in browser. Sure seems to do something as per docs. So please suggest for a way out as this is critical to our app.
 
Screenshots attached below to indicate the heap memory in a few seconds and the corresponding console.logs mentioned in point 4

![Screenshot from 2020-11-23 17-48-24](https://user-images.githubusercontent.com/11805936/99961437-71b54380-2db4-11eb-9ffb-e3346e56b108.png)
![Screenshot from 2020-11-23 17-39-46](https://user-images.githubusercontent.com/11805936/99961452-7a0d7e80-2db4-11eb-8230-bd6e4e324cec.png)

## System information

- OS: checked on macOS, ubuntu
- Version of Next.js: 9.5
- Version of Node.js: 10.19
- Deployment(LocalHost ) `nodemon --watch app --inspect --watch app/server.js`

## Additional context

This issue is bugging us in production as well since months now. We are handling memory overload via restart till threshold.
Interestingly this issue only occurs and logs all the repeated console.logs when node devtools are open. Otherwise, there is no such repetition of console.logs. However, app does grow out of memory as observed on local machines and other servers where app is deployed.

**Comment by timneutkens** [comment]
```
server.get('*', async (req, res) => {
        console.log(req.url, 'render url');
        return handler(req, res);
     });
```

you added a console.log inside of server.get, meaning every single request to the server logs a line, this has nothing to do with `app.prepare` re-executing (it doesn't). It seems like `/json` and `/json/version` are being called continuously, maybe in your application?

**Reply by kvedantmahajan** [reply]
Umm!!   Not sure what would do that, but it only happens when I open the node dev tools while the app is running on localhost and not normally on `npm run start:dev` i.e. `nodemon --watch app --inspect --watch app/server.js` command. The app otherwise serves fine, with only one entry of `/json` in terminal. 
@timneutkens
